refactor(core): update peerDependencies to allow rxjs7

crisbeto opened the pull request:
We can't update the framework to rxjs7 until version 13, because it contains breaking changes, but we can allow users to opt into it since all of our code should be compatible.

These changes expand the allowed version range of rxjs and add an integration test to verify that we don't get compilation errors. Note that we also have a test that runs the AIO examples against rxjs 7 already (#42660).

Fixes #41897.

gkalpak commented on packages/common/test/pipes/async_pipe_spec.ts:
```diff
@@ -22,7 +22,7 @@ import {Subscribable, Unsubscribable} from 'rxjs';
       const wrapSubscribable = <T>(input: Subscribable<T>): Subscribable<T> => ({
         subscribe(...args: any): Unsubscribable {
-          const subscription = input.subscribe(...args);
+          const subscription = input.subscribe.apply(input, args);
```
OOC, why is this change necessary?

crisbeto replied:
We'll have to make it when we eventually switch to rxjs 7. I figured I'd do it now since it's in a test.

gkalpak replied:
Yeah, but why is this change necessary for RxJS v7? Aren't `input.subscribe(...args)` and `input.subscribe.apply(input, args)` equivalent?

crisbeto replied:
I honestly don't remember since it's been a while from when I tried updating it. I just know that it was a compilation error.

gkalpak replied:
I was curious, so I tried it out. The error is:
`A spread argument must either have a tuple type or be passed to a rest parameter.`

Alternatively, this could be fixed by changing the type of `args` (but `.apply()` is fine as well, I guess):

```diff
-        subscribe(...args: any): Unsubscribable {
+        subscribe(...args: Parameters<Subscribable<T>['subscribe']>): Unsubscribable {
           const subscription = input.subscribe(...args);
```

gkalpak commented on integration/typings_test_rxjs7/package.json:
```diff
@@ -0,0 +1,28 @@
+    "@angular/upgrade": "file:../../dist/packages-dist/upgrade",
+    "@types/jasmine": "file:../../node_modules/@types/jasmine",
+    "rxjs": "^7.3.0",
+    "typescript": "~4.3.4",
```
Why not `file:../../node_modules/typescript`?

gkalpak commented on integration/typings_test_rxjs7/tsconfig.json:
```diff
@@ -0,0 +1,26 @@
+  },
+  "files": [
+    "include-all.ts",
+    "node_modules/@types/jasmine/index.d.ts"
```
OOC, shouldn't this be in the `types` list?

crisbeto replied:
Not sure, I was following the same pattern as the TS tests.

gkalpak reviewed (approved):
Not sure if `feat` is the right commit type. Is that what we've uses for other similar changes in the past?
Otherwise lgtm :+1:

AndrewKushnir commented on packages/core/test/acceptance/property_binding_spec.ts:
```diff
@@ -187,7 +187,7 @@ describe('property bindings', () => {
     class OtherDir {
       @Input() id: number|undefined;
-      @Output('click') clickStream = new EventEmitter();
+      @Output('click') clickStream = new EventEmitter<void>();
```
Just curious how break-y that would be for apps code as well and whether we should consider doing a migration (schematic) to simplify the upgrade? That also reminds me of #38916...

crisbeto replied:
It'll be pretty break-ey, even for our own code, but as you mentioned, it's not a problem for now because version 7 is opt in. I explored it a bit in https://github.com/angular/angular/pull/41793 which revealed a few other breaking changes.

AndrewKushnir reviewed (approved):
LGTM, thanks @crisbeto 👍 

I've just left a comment about `EventEmitter`. This is not a problem for now since RxJS v7 is an opt-in, but we'd probably need to look at it when we decide to drop RxJS v6 support (in favor of v7).

IgorMinar reviewed (approved):
This looks good to me. With all the aio tests running on rxjs 7 would should have enough test coverage to know when things break.

It's going to be really interesting to see how easy it will be for everyone to get from 6 to 7 and collect pain points that we might need to help or resolve before removing v6 support.

Thanks for getting this thought the line.

Reviewed-for: global-approvers

crisbeto committed:
refactor(core): update peerDependencies to allow rxjs7

We can't update the framework to rxjs7 until version 13, because it contains breaking changes, but we can allow users to opt into it since all of our code should be compatible.

These changes expand the allowed version range of rxjs and add an integration test to verify that we don't get compilation errors. Note that we also have a test that runs the AIO examples against rxjs 7 already (#42660).

Fixes #41897.
